hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/SDKStreamDrainer.java

```diff
@@ -34,20 +32,17 @@
 import static java.util.Objects.requireNonNull;
 import static org.apache.hadoop.fs.s3a.impl.InternalConstants.DRAIN_BUFFER_SIZE;
 import static org.apache.hadoop.fs.statistics.impl.IOStatisticsBinding.invokeTrackingDuration;
-import static org.apache.hadoop.io.IOUtils.cleanupWithLogger;
 
 /**
  * Drains/aborts s3 or other AWS SDK streams.
  * It is callable so can be passed directly to a submitter
  * for async invocation.
- * A request object may be passed in; it will be implicitly
- * cached until this object is GCd.
- * This is because in some versions of the AWS SDK, the S3Object
- * has a finalize() method which releases the http connection,
- * even when the stream is still open.
- * See HADOOP-17338 for details.
  */
-public class SDKStreamDrainer implements CallableRaisingIOE<Boolean> {
+// TODO:WiP: the input stream needs to also be Abortable and
+//  v2 does not seem to have a super-class like SdkFilterInputStream.
+// * use generics or resort to runtime cast?
+public class SDKStreamDrainer<TStream extends InputStream & Abortable>
```

Comment:
`ResponseInputStream` has an abort method, can we not use that?

Reply:
That should work, but would make the tests more difficult since `ResponseInputStream` is final - see `FakeSDKInputStream`.